[Alex Teo] iP

alexteo98 opened the pull request:
# Bro
> What did I need to do again? - <s>You</s> Me, <b>everyday</b> 
## Ever forgot what you needed to do? :frowning_face: Now you won't! :relieved:

With <b>Bro</b>, you now have an easy way to remember tasks, deadlines and important events. 
It is:
* Easy to use
* Text based

How to start:
1. [Click here](https://github.com/alexteo98/ip/releases/download/A-Release/Bro.jar) to download
2. Double click on "Bro.jar" to start!

Features supported:
- [x] Add/Remove Tasks
- [x] Find Tasks
- [x] Sort Tasks
- [x] Graphical User Interface 


If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

alexteo98 committed:
Echo and Exit working.
Level 1 completed.

alexteo98 committed:
Adding task done, list tasks done.
Level 2 completed.

alexteo98 committed:
Added javadocs for Duke methods

alexteo98 committed:
Added new TaskList Data structure.
Mark working.

alexteo98 committed:
unmark working.
Level 3 completed.

alexteo98 committed:
Removed Message interface.
Added todo,events,deadlines sub classes
Level 4 completed.

alexteo98 committed:
Removed Message interface.
Added todo,events,deadlines sub classes
Level 4 completed.

alexteo98 committed:
Formatted output of chatbot.
Verified tests.
A-TextUiTesting completed.

alexteo98 committed:
Created exceptions for invalid date/time,description,type

alexteo98 committed:
Marking index out of bounds error solved.

alexteo98 committed:
Invalid Task description and date/time implemented for deadline

alexteo98 committed:
Fixed bug when task name includes mark/unmark
Repeated marking/unmarking fixed.
Marking tasks out of bounds fixed.
Adding invalid task type fixed.
Invalid date/time/description added for deadline,todo and event.

alexteo98 committed:
Changed tasklist to taskmanager.
Changed implementation of mark/unmark in taskmanager.
Added delete task functionality.
Changed program flow in main function.
Level 6 completed.

alexteo98 committed:
Added help menu on invalid entry

alexteo98 committed:
Added new task function that was previously removed

alexteo98 committed:
Revert "Added new task function that was previously removed"

This reverts commit 152a8e7d7689e4e59e3448705c4a4ce4c2340ada.

alexteo98 committed:
Basic file operations complete. Errors not handled completely

alexteo98 committed:
Basic save/load done.
Parse saved string to task object added.
Level-7 completed.

alexteo98 committed:
Date/Time support implemented
Level-8 completed.

alexteo98 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Basic save/load done. Parse saved string to task object added. Level-7 completed.
  Basic file operations complete. Errors not handled completely

alexteo98 committed:
Merge branch 'branch-level-8'

* branch-level-8:
  Date/Time support implemented Level-8 completed.

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Task.java

alexteo98 committed:
Solved merge conflicts.
Saving supported with date/time

alexteo98 committed:
Changed data file paths to ~/Duke/data, OS independent

alexteo98 committed:
Changed return types in TaskManager. Modified error messages on parsing error.

alexteo98 committed:
Abstracted UI and updated Duke program flow

alexteo98 committed:
Add Storage, Parser classes

Methods dealing with storage like saving and loading are in Duke class.
Parsing of input from user and file to Task objects are handled by Task and
TaskManager class.

Common methods dealing with files and parsing of input should be abstracted
into their own classes. Having their own classes allows easier manipulation
through their respective public methods.

Let's extract all methods related to storage and parsing into a seperate class.

Minor renaming and refactoring are also included in this commit.

alexteo98 committed:
Fix ArrayOutOfBounds error in Parser

alexteo98 committed:
Update Duke to use Storage class

alexteo98 committed:
Add Command class

Commands entered by User is parsed by main Duke class.

Extracting the various commands into a seperate Command class
will reduce spagetti code and provides the use of inheritance
for various command types.

Let's:
* Create A new class that represents the various actions that
are available to the user.

This will allow the use of polymorphism when executing commands.

* Note: only skeletal code is added for now, implementations
will follow in future commits.
* Parser for commands is also added.
* Main Duke class does not use Command class yet, will refactor
it to use Commands class once it is functional.
* Minor changes in Storage and Ui are also included in this commit.

alexteo98 committed:
Update parsing functions to Parser Class

alexteo98 committed:
Update Ui display methods to non-static

alexteo98 committed:
Update Duke and Command class to use ui instance methods

alexteo98 committed:
Add Unit Tests

alexteo98 committed:
Add more tests to ParserTest

alexteo98 committed:
Add functionality to Command Class

alexteo98 committed:
Update Duke to use Command Class

alexteo98 committed:
Refactor Duke into packages

alexteo98 committed:
Add support for text based dates

alexteo98 committed:
Add feature sort task list by name and date

alexteo98 committed:
Add overloaded methods for Task Manager

alexteo98 committed:
Document Command classes

alexteo98 committed:
Document Task classes

alexteo98 committed:
Fix bug when parsing date string from file

alexteo98 committed:
Fix bug when parsing date string from file

alexteo98 committed:
Add feature to search for tasks

alexteo98 committed:
Add feature to search for tasks

alexteo98 committed:
Update to follow coding standard

alexteo98 committed:
Merge commit '7ad2fc4dd4589f05d6c47547d60d30b5ed92b086'

alexteo98 committed:
Merge branch 'branch-Level-9'

alexteo98 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/TaskManager.java
#	src/main/java/duke/commands/AddCommand.java
#	src/main/java/duke/commands/Command.java
#	src/main/java/duke/commands/DeleteTaskCommand.java
#	src/main/java/duke/commands/ExitCommand.java
#	src/main/java/duke/commands/ListCommand.java
#	src/main/java/duke/commands/MarkCommand.java
#	src/main/java/duke/commands/MarkDoneCommand.java
#	src/main/java/duke/commands/SortByDateCommand.java
#	src/main/java/duke/commands/SortByNameCommand.java
#	src/main/java/duke/tasks/Deadline.java
#	src/main/java/duke/tasks/Event.java
#	src/main/java/duke/tasks/Task.java

alexteo98 committed:
Refactor saving format to Parser.class

* Move functions for task saving to Parser.class.
* Move task marking checks to Task.class.
* Remove old unused exceptions. (+1 squashed commits)

Squashed commits:

[0d16451] Refactor saving format to Parser.class

* Move functions for task saving to Parser.class.
* Move task marking checks to Task.class.
* Remove old unused exceptions.

alexteo98 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

alexteo98 committed:
Add gradle support for build, test, and JavaDoc

alexteo98 committed:
Merge branch 'branch-A-Gradle'

alexteo98 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Parser.java

alexteo98 committed:
Add CheckStyle

alexteo98 committed:
Merge branch 'branch-A-CheckStyle'

Graphcalibur commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,201 @@
+
+
+        switch(type) {
+        case 'T':
```
Good job on following the coding standard with regards to indentation for switch and case!

Graphcalibur commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,86 @@
+        File dir = new File(dirPath);
+        return dir.exists();
+    }
+    public boolean createDir() {
```
I think there should be a line break before here to separate the two functions.

Graphcalibur commented on src/main/java/duke/TaskManager.java:
```diff
@@ -0,0 +1,96 @@
+        this.tasks = tasks;
+    }
+
+    public void addTask(Task t) {
```
These methods are missing their javadoc header comments. If you have time, I'd recommend adding them because it really makes the code easier to read.

Graphcalibur commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,201 @@
+            return null;
+        }
+    }
+    private static Todo newToDo(String s) throws DukeException {
```
I think it would be more consistent to have these newX() method names to be phrased as verbs.

Graphcalibur commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,208 @@
+
+public class Ui {
+
+    private static final String line =
```
These constant variables should have all uppercase names to differentiate them.

Graphcalibur reviewed (commented):
Overall, good job! You mostly followed the coding standard practices with just a few hiccups I've highlighted.

seanleong339 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,201 @@
+import duke.tasks.Task;
+import duke.tasks.Todo;
+
+public class Parser {
```
Perhaps include a header comment for this class? 😄 
I noticed this in some other classes as well.

seanleong339 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,201 @@
+        return String.format("%c\t%c\t%s\t%s", t.getType(), t.getDone(), t.getTaskName(), dateStr);
+    }
+
+    public static Task parseToTaskFromFile(String fileInput) throws DukeException {
```
Good that you are specific with your method name! 👍

seanleong339 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,86 @@
+            throw new DukeException("Unable to read Task List from file!");
+        }
+    }
+    public boolean saveTaskManager(TaskManager taskManager) throws DukeException {
```
Perhaps add a line break here to make the code neater?

seanleong339 commented on src/main/java/duke/exceptions/FailedTaskParseException.java:
```diff
@@ -0,0 +1,7 @@
+package duke.exceptions;
+
+public class FailedTaskParseException extends DukeException {
```
Perhaps add a header comment to explain when this exception occurs or is used?

seanleong339 reviewed (commented):
Overall, good naming for the variables! They were very specific and helped me to understand the code

alexteo98 committed:
Add Gui to Duke

alexteo98 committed:
Add Starting messages to Duke

alexteo98 committed:
Remove extra Launcher Class

alexteo98 committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	build.gradle
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/commands/AddCommand.java
#	src/main/java/duke/commands/DeleteTaskCommand.java
#	src/main/java/duke/commands/FindCommand.java
#	src/main/java/duke/commands/MarkDoneCommand.java
#	src/main/java/duke/commands/MarkUndoneCommand.java

alexteo98 committed:
Fix Ui to auto-scroll and disable resize of window

alexteo98 committed:
Add Update feature to Duke

Users were unable to change the date or name of the task after adding.

Lets:
* Create a new command to update the name and date of the tasks

alexteo98 committed:
Merge branch 'branch-BCD-Extension'

alexteo98 committed:
Change behaviour of list date and list name in Commands

alexteo98 committed:
Add exit behaviour for GUI

alexteo98 committed:
Refactor Command parsing to Parser class

Parsing logic for each type of command is handled by the respective
command classes.

This makes it difficult to create instances of the commands.

Extracting the parsing logic of these commands into Parser class
makes the parser class the only class responsible for understanding
user input.

Let's:
* Remove all logic dealing with user input in each command class and
put them in parser class.
* Create the fields needed by each command as private variables.

Dead code and classes are also removed in this commit.

alexteo98 committed:
Add assertions to commands and Duke

Important assumptions during program flow are not enforced. By adding
assertions, we can ensure these assumptions are held true.

Lets:
* Add assertions to main to ensure program has not ended.
* Add assertions to command parsing to ensure a command is able to be
executed.
* Add assertions to important variables needed for command execution.

alexteo98 committed:
Follow Code Quality standards and CheckStyle

alexteo98 committed:
Merge pull request #3 from alexteo98/branch-A-CodeQuality

Follow Code Quality standards and CheckStyle

alexteo98 committed:
Merge pull request #2 from alexteo98/branch-A-Assertions

Add assertions to commands and Duke

alexteo98 committed:
Update Duke to new GUI

alexteo98 committed:
Add screenshot to docs

alexteo98 committed:
Add user guide

alexteo98 committed:
Update README.md

alexteo98 committed:
Update README for HTML

alexteo98 committed:
Update README.md

alexteo98 committed:
Update README.md

alexteo98 committed:
Merge pull request #4 from alexteo98/alexteo98-patch-2

Alexteo98 patch 2

alexteo98 committed:
Update Duke to Bro
